filesystems/drs/src/main/scala/cromwell/filesystems/drs/DrsPathBuilderFactory.scala

```diff
@@ -39,37 +41,55 @@ class DrsPathBuilderFactory(globalConfig: Config, instanceConfig: Config, single
     val storage = StorageOptions.newBuilder().setCredentials(credentials).build().getService
 
     IO.delay {
       val blob = storage.get(gcsFile.bucket, gcsFile.file)
       blob.reader()
+    } handleErrorWith { throwable =>
+      (requesterPaysProjectIdOption, throwable) match {
+        case (Some(requesterPaysProjectId), storageException: StorageException)
+          if storageException.getMessage == "Bucket is requester pays bucket but no user project provided." =>
+          IO.delay {
+            val blob = storage.get(gcsFile.bucket, gcsFile.file, BlobGetOption.userProject(requesterPaysProjectId))
+            blob.reader(Blob.BlobSourceOption.userProject(requesterPaysProjectId))
+          }
+        case _ => IO.raiseError(throwable)
+      }
     }
   }
 
 
-  private def inputReadChannel(url: String, urlScheme: String, serviceAccount: String): IO[ReadableByteChannel] =  {
+  private def inputReadChannel(url: String,
+                               urlScheme: String,
+                               serviceAccountJson: String,
+                               requesterPaysProjectIdOption: Option[String]): IO[ReadableByteChannel] =  {
     urlScheme match {
       case GcsScheme =>
         val Array(bucket, fileToBeLocalized) = url.replace(s"$GcsScheme://", "").split("/", 2)
-        gcsInputStream(GcsFilePath(bucket, fileToBeLocalized), serviceAccount)
+        gcsInputStream(GcsFilePath(bucket, fileToBeLocalized), serviceAccountJson, requesterPaysProjectIdOption)
       case otherScheme => IO.raiseError(new UnsupportedOperationException(s"DRS currently doesn't support reading files for $otherScheme."))
     }
   }
 
 
-  private def drsReadInterpreter(marthaResponse: MarthaResponse): IO[ReadableByteChannel] = {
-    val serviceAccount = marthaResponse.googleServiceAccount match {
-      case Some(googleSA) => googleSA.data.toString
-      case None => throw new GoogleSANotFoundException
+  private def drsReadInterpreter(options: WorkflowOptions, requesterPaysProjectIdOption: Option[String])
+                                (marthaResponse: MarthaResponse): IO[ReadableByteChannel] = {
+    val serviceAccountJsonIo = marthaResponse.googleServiceAccount match {
+      case Some(googleSA) => IO.pure(googleSA.data.toString)
+      case None => IO.fromEither(options.get(GoogleAuthMode.UserServiceAccountKey).toEither)
     }
 
     //Currently, Martha only supports resolving DRS paths to GCS paths
-    DrsResolver.extractUrlForScheme(marthaResponse.dos.data_object.urls, GcsScheme) match {
-      case Right(url) => inputReadChannel(url, GcsScheme, serviceAccount)
-      case Left(e) => IO.raiseError(e)
+    serviceAccountJsonIo flatMap { serviceAccountJson =>
+      DrsResolver.extractUrlForScheme(marthaResponse.dos.data_object.urls, GcsScheme) match {
+        case Right(url) => inputReadChannel(url, GcsScheme, serviceAccountJson, requesterPaysProjectIdOption)
+        case Left(e) => IO.raiseError(e)
```

Comment:
Is there a reason we're pattern matching then throwing rather than just using a `flatMap` here?

Reply:
Refactored it to for-comprehension